File: Exercises_Java/Alarm_System/src/impl/SmokeSensor.java
package impl;

import java.util.Random;

public class SmokeSensor implements HazardSensor {

  private String location;
  private String sensorType;
  private double batteryPercentage;
  private Random randomint;

  public SmokeSensor(String location) {
    this.location = location;
    this.sensorType = "Smoke_sensor";
    this.batteryPercentage = 100;
    randomint = new Random();
    //randomint.setSeed(2);
  }

  /**
   * Raises an alarm 10% of the time it is called, and drains 20% battery between each poll
   *
   * @return true 10% of the time
   * Assume that the sensor can't be triggered if the battery percentage is zero
   */
  @Override
  public boolean isTriggered() {
    if (this.getBatteryPercentage() == 0) return false;
    else {
      batteryPercentage -= 20;
      return randomint.nextInt(100) <= 10;
    }
  }

  @Override
  public String getLocation() {
    return location;
  }

  @Override
  public String getSensorType() {
    return sensorType;
  }

  @Override
  public double getBatteryPercentage() {
    return batteryPercentage;
  }

}

File: Exercises_Scala/week_02/src/src/impl/SmokeSensor.scala
package impl

/**
  * Created by lucieburgess on 26/01/2017.
  */

import scala.util.Random

class SmokeSensor(location: String) extends Sensor with HazardSensor {

  /**
    * created by LucieCBurgess on 24.01.2017
    * Task is to implement and test the SmokeSensor methods
    * For this I created a mew trait called HazardSensor. Both SmokeSensor and FireSensor implement interface HazardSensor
    */

    private val sensorType: String = "Smoke_sensor"
    private var batteryPercentage: Int = 100
    private var randomint = new Random

    override def isTriggered: Boolean = {
      if (getBatteryPercentage != 0) {
        batteryPercentage -= 20
      }
      return Random.nextInt() < 0.1
    }

    override def getLocation: String = location

    override def getSensorType: String = sensorType

    override def getBatteryPercentage: Double = batteryPercentage

}
